### PR TITLE
pkg/trace: turn into a stand-alone module

### PR DESCRIPTION
**Pull request by gbbr**
This change creates a module out of `pkg/trace` in order for this to be used by other external modules when processing traces. It was necessary to break all dependencies on `datadog-agent` in order to avoid pulling it.

⚠️ Each commit removes one dependency. Please review on a commit by commit basis.

### QA Plan

In order for correct testing, any system which was touched (in order for it to be moved out) must be tested independently. The list is not exhaustive, but should contain the major components:

* Performance testing to ensure that additional wrappers and interfaces that were introduced do not cause any degrade.
* All log levels should work and output messages.
* AppSec proxy continues to work efficiently despite its EOL period.
* All config continues to be registered correctly by the command. Extra testing that the below functions is necessary:
    * Windows Pipes
    * GUI Port gets picked up.
    * Dogstatsd settings get picked up (including socket).
    * Proxy settings for: profiling, AppSec, Debugger.
    * MaxCatalogEntries.
* Global Proxy settings continue to work.
* Config endpoint works (behind `config_endpoint` feature flag).
* Remote rates work (behind `remote_rates` feature flag).
* Internal profiling works.
* Tagger (container tags) work.

### Follow-up work needed:

* Move out the `info` package version-related stuff. It makes no sense in the module. It should be set as part of the configuration of the agent and in the context of the `trace-agent` command have a single source of truth: `agent.AgentVersion`
* Make `internal` all packages that are not needed.
* Make follow up changes based on usage elsewhere.

### CODE REVIEW

**Review by raphaelgavache — APPROVED** [review]
lgtm

File: pkg/trace/watchdog/logonpanic_test.go [hunk 1]
```diff
@@ -24,11 +24,7 @@ func init() {
 	if err != nil {
 		panic(err)
 	}
-	err = seelog.ReplaceLogger(logger)
-	if err != nil {
-		panic(err)
-	}
-	log.SetupLogger(logger, "INFO")
+	log.SetLogger(logger)
```
Comment (mx-psi) [inline pkg/trace/watchdog/logonpanic_test.go:27]: Would be nice to get rid of `seelog` entirely by using a mock logger here instead (maybe not very important, it just would be nice to not depend on it)

File: pkg/trace/config/client.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	"net/http"
+	"sync"
+	"time"
+)
+
+// TODO(gbbr): Perhaps this is not the best place for this structure.
+
+// ResetClient wraps (http.Client).Do and resets the underlying connections at the
+// configured interval
+type ResetClient struct {
```
Comment (mx-psi) [inline pkg/trace/config/client.go:18]: Is this copied over from another part of the codebase? If so, should we add a note about it?

**Review by mx-psi — COMMENTED** [review]
This is great, and I am sure it took a lot of work 🙇 🙇  I think overall it looks good, I left a few inline comments, and have some other small comments that I couldn't make inline to polish this:
- You can move `cmd/trace-agent/flags` and `cmd/trace-agent/osutil` to `cmd/trace-agent/internal/flags` and `cmd/trace-agent/internal/osutil` to make them [internal](https://go.dev/doc/go1.4#internalpackages). There are other packages that you could make internal on `pkg/trace`, but for those we need to know what parts of the API we will need on the Datadog exporter, so we can leave those for later.
- The `pkg/trace/info` package still has a dependency of sorts with the Agent codebase: its `Version`, `GitCommit` and related vars are set at build time with the `make.go`. I would suggest at least isolating the parts that are set at build time into a new package (and if it's at all possible to move it out of the module, I think that would be best)

**Comment by mx-psi** [comment]
Also, I think this should be labeled as a major change, and we should write QA instructions for this. It touches a lot of things and the risk of breaking something is high

**Review by mx-psi — COMMENTED** [review]
One more thing (and there will probably be even more that I think about, sorry 😄): It would be nice to have a `doc.go` like this one https://github.com/DataDog/datadog-agent/blob/main/pkg/otlp/model/doc.go describing the purpose of the module and explicitly saying that the API is unstable and may have breaking changes

File: pkg/trace/go.mod [hunk 3]
```diff
@@ -0,0 +1,30 @@
+module github.com/DataDog/datadog-agent/pkg/trace
+
+go 1.16
+
+require (
+	github.com/DataDog/datadog-agent/pkg/obfuscate v0.34.0
+	github.com/DataDog/datadog-go v4.8.3+incompatible
+	github.com/DataDog/sketches-go v1.2.1
+	github.com/DataDog/viper v1.9.0 // indirect
```
[the remark this answers is not in the document]
Comment (gbbr) [reply pkg/trace/go.mod:9]: What do you mean? I've not added it myself. It is an indirect dependency.

[the remark this answers is not in the document]
Comment (mx-psi) [reply pkg/trace/go.mod]: It's not really an indirect dependency, but Go added it for some reason 🤷 I have seen this happen before, Go won't remove an `// indirect` dependency when tidying even if it's not actually used

**Comment by mx-psi** [comment]
You probably need to do `inv -e tidy-all`

**Comment by gbbr** [comment]
@mx-psi I think I should've addressed all comments, and have added a brief QA plan. Please take another look once you get a chance.

**Review by maxday — APPROVED** [review]

**Review by mx-psi — APPROVED** [review]
:rocket:

**Comment by gbbr** [comment]
This change creates a module out of `pkg/trace` in order for this to be used by other external modules when processing traces. It was necessary to break all dependencies on `datadog-agent` in order to avoid pulling it.

⚠️ Each commit removes one dependency. Please review on a commit by commit basis.

### QA Plan

In order for correct testing, any system which was touched (in order for it to be moved out) must be tested independently. The list is not exhaustive, but should contain the major components:

* Performance testing to ensure that additional wrappers and interfaces that were introduced do not cause any degrade.
* All log levels should work and output messages.
* AppSec proxy continues to work efficiently despite its EOL period.
* All config continues to be registered correctly by the command. Extra testing that the below functions is necessary:
    * Windows Pipes
    * GUI Port gets picked up.
    * Dogstatsd settings get picked up (including socket).
    * Proxy settings for: profiling, AppSec, Debugger.
    * MaxCatalogEntries.
* Global Proxy settings continue to work.
* Config endpoint works (behind `config_endpoint` feature flag).
* Remote rates work (behind `remote_rates` feature flag).
* Internal profiling works.
* Tagger (container tags) work.

### Follow-up work needed:

* Move out the `info` package version-related stuff. It makes no sense in the module. It should be set as part of the configuration of the agent and in the context of the `trace-agent` command have a single source of truth: `agent.AgentVersion`
* Make `internal` all packages that are not needed.
* Make follow up changes based on usage elsewhere.

**Pull request merged**